src/Npgsql/TypeHandlers/ArrayHandler.cs

```diff
@@ -23,85 +25,69 @@ namespace Npgsql.TypeHandlers
+            LowerBound = lowerBound;
+            ElementHandler = elementHandler;
         }
 
+        internal override Type GetFieldType(FieldDescription? fieldDescription = null) => typeof(Array);
+        internal override Type GetProviderSpecificFieldType(FieldDescription? fieldDescription = null) => typeof(Array);
+
         /// <inheritdoc />
         public override ArrayHandler CreateArrayHandler(PostgresArrayType arrayBackendType)
             => throw new NotSupportedException();
 
         /// <inheritdoc />
         public override RangeHandler CreateRangeHandler(PostgresRangeType rangeBackendType)
             => throw new NotSupportedException();
-    }
-
-    /// <summary>
-    /// Base class for all type handlers which handle PostgreSQL arrays.
-    /// </summary>
-    /// <remarks>
-    /// http://www.postgresql.org/docs/current/static/arrays.html.
-    ///
-    /// The type handler API allows customizing Npgsql's behavior in powerful ways. However, although it is public, it
-    /// should be considered somewhat unstable, and  may change in breaking ways, including in non-major releases.
-    /// Use it at your own risk.
-    /// </remarks>
-    public class ArrayHandler<TElement> : ArrayHandler
-    {
-        readonly int _lowerBound; // The lower bound value sent to the backend when writing arrays. Normally 1 (the PG default) but is 0 for OIDVector.
-        readonly NpgsqlTypeHandler _elementHandler;
-
-        /// <inheritdoc />
-        public ArrayHandler(PostgresType arrayPostgresType, NpgsqlTypeHandler elementHandler, int lowerBound = 1)
-            : base(arrayPostgresType)
-        {
-            _lowerBound = lowerBound;
-            _elementHandler = elementHandler;
-        }
-
-        internal override Type GetFieldType(FieldDescription? fieldDescription = null) => typeof(Array);
-        internal override Type GetProviderSpecificFieldType(FieldDescription? fieldDescription = null) => typeof(Array);
 
         #region Read
 
-        internal override TAny Read<TAny>(NpgsqlReadBuffer buf, int len, FieldDescription? fieldDescription = null)
-            => Read<TAny>(buf, len, false, fieldDescription).Result;
+        internal override TRequestedArray Read<TRequestedArray>(NpgsqlReadBuffer buf, int len, FieldDescription? fieldDescription = null)
+            => Read<TRequestedArray>(buf, len, false, fieldDescription).GetAwaiter().GetResult();
 
         /// <inheritdoc />
-        protected internal override async ValueTask<TAny> Read<TAny>(NpgsqlReadBuffer buf, int len, bool async, FieldDescription? fieldDescription = null)
+        protected internal override async ValueTask<TRequestedArray> Read<TRequestedArray>(NpgsqlReadBuffer buf, int len, bool async, FieldDescription? fieldDescription = null)
         {
-            if (IsArrayOf<TAny, TElement>.Value)
-                return (TAny)(object)await ReadArray<TElement>(buf, async);
+            Debug.Assert(ArrayTypeInfo<TRequestedArray>.IsArrayOrList);
```

Comment:
You're probably right.
I'd consider the exception as vital, because `ArrayHandler` is on the public API surface, so this method might be called in a way that we don't expect or support but in this case the `Debug.Assert` is really not necessary.

I think it's a leftover from some phase during refactoring when the method body was more complicated and we felt like we'd need it to make clear what the method expects.
I'll remove it.